### PR TITLE
Add transaction dictionary, to replace the object current state vars.

### PR DESCRIPTION
**Pull request by yaacov**
Add transaction support for transaction id field in the ModbusTCP packet.

Issue: https://github.com/yaacov/node-modbus-serial/issues/31

### CODE REVIEW

**Comment by yaacov** [comment]
@valfis86 can you check if this fixes your problem ?
@meteocontrol can you look at it ?

**Comment by connium** [comment]
Sure, will check the code with @meteocontrol and @DLoT tomorrow.

**Comment by yaacov** [comment]
@connium thanks :-)

**Comment by yaacov** [comment]
@valfis86 ayt ?

File: index.js [hunk 1]
```diff
@@ -124,8 +124,10 @@ function _readFC16(data, next) {
  * @private
  */
 function _writeBufferToPort(buffer) {
+    var transaction = this._transactions[this._transactionId]
+
     this._port.write(buffer);
-    this._timeoutHandle = _startTimeout(this._timeout, this._next);
+    this._timeoutHandle = _startTimeout(this._timeout, transaction.next);
```
Comment (connium) [inline index.js]: Each request needs to have its own timeout.

**Example**
Think of a timeout of 5 seconds.
* 0:00.00: send request to device 1 (transaction id TID = 1)
* 0:00.03: send request to device 2 (transaction id TID = 2)
  * overwrite timeout handle
* 0:00.04: receive response of TID 1
  * cancel timeout handle (which is the timeout handle of TID 2)
* 0:00.05: receive response of TID 2
  * no timeout handle to cancel
* 0:00.005: emit timeout for TID 1

**Comment by connium** [comment]
Beside the above mentioned issue, the code seems to work (we cannot test Modbus TCP yet).

In general I think the code should be split into multiple, smaller modules which will help to organize and maintain the code. We will open some PRs in the near future to address this, if you are fine with that.

**Comment by yaacov** [comment]
> We will open some PRs in the near future to address this

I'm always happy for help :-)
p.s. @valfis86 is not responding , and I too can't test this feature in the wild.

File: index.js (hunk 1 above)
Comment (yaacov) [inline index.js]: @connium :+1: fixed

**Comment by yaacov** [comment]
@connium @valfis86 I am ready to merge this, but I do not know if it solves the original issue, any notes ?

**Comment by valfis86** [comment]
Sorry, didn't have any time to verify. I had just tested different modbus
libraries, but the project became obsolete.

2016-07-08 10:07 GMT+02:00 Yaacov Zamir <notifications@github.com>:

**Comment by yaacov** [comment]
@valfis86 thanks for the replay, we were worried about you :-)

**Pull request merged**